src/Derivatives/forward.jl

```diff
@@ -1,3 +1,15 @@
+# Internal function: evaluates `func` at `args` and checks that the return type
+# is `RetType`. This is a catch to provide the user with a nice error message if
+# their user-defined function returns unexpected results.
+function eval_and_check_return_type(func::Function, RetType, args...)
+    ret = func(args...)
+    if !isa(ret, RetType)
+        error("Expected return type of $(RetType), but got $(typeof(ret)). " *
+              "Make sure your user-defined function only depends on variables" *
```

Comment:
This message seems too specialized. Maybe include it only if `ret` is an `AbstractJuMPScalar`?

Reply:
Does `using ..JuMP` work in the derivatives module?